pulsar-io/elastic-search/src/main/java/org/apache/pulsar/io/elasticsearch/client/RestClientFactory.java

```diff
@@ -0,0 +1,105 @@
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.pulsar.io.elasticsearch.client;
+
+
+import com.fasterxml.jackson.databind.ObjectMapper;
+import lombok.extern.slf4j.Slf4j;
+import org.apache.commons.lang3.StringUtils;
+import org.apache.http.client.methods.HttpGet;
+import org.apache.pulsar.io.elasticsearch.ElasticSearchConfig;
+import org.apache.pulsar.io.elasticsearch.client.elastic.ElasticSearchJavaRestClient;
+import org.apache.pulsar.io.elasticsearch.client.opensearch.OpenSearchHighLevelRestClient;
+import org.opensearch.client.Request;
+import org.opensearch.client.Response;
+
+import java.io.IOException;
+import java.net.MalformedURLException;
+import java.util.Map;
+
+@Slf4j
+public class RestClientFactory {
+
+    private static final ObjectMapper MAPPER = new ObjectMapper();
+
+    public static RestClient createClient(ElasticSearchConfig config, BulkProcessor.Listener bulkListener) throws MalformedURLException {
+
+        if (config.getCompatibilityMode() == ElasticSearchConfig.CompatibilityMode.ELASTICSEARCH) {
+            log.info("Found compatibilityMode set to '{}', using the ElasticSearch Java client.", config.getCompatibilityMode());
+            return new ElasticSearchJavaRestClient(config, bulkListener);
+        } else if (config.getCompatibilityMode() == ElasticSearchConfig.CompatibilityMode.ELASTICSEARCH_7 ||
+                config.getCompatibilityMode() == ElasticSearchConfig.CompatibilityMode.OPENSEARCH) {
+            log.info("Found compatibilityMode set to '{}', using the OpenSearch High Level Rest API Client.", config.getCompatibilityMode());
+            return new OpenSearchHighLevelRestClient(config, bulkListener);
+        }
+        log.info("Found compatibilityMode set to '{}', will try to auto detect the best client to use.", config.getCompatibilityMode());
+        try {
+            final Map<String, Object> jsonResponse = requestInfo(config);
+            final boolean useOpenSearchHighLevelClient = useOpenSearchHighLevelClient(jsonResponse);
+            log.info("useOpenSearchHighLevelClient={}, got info response: {}", useOpenSearchHighLevelClient, jsonResponse);
+            if (useOpenSearchHighLevelClient) {
+                return new OpenSearchHighLevelRestClient(config, bulkListener);
+            }
+            return new ElasticSearchJavaRestClient(config, bulkListener);
+        } catch (IOException ioException) {
+            log.trace("Got error while performing info request: {}, falling back to OpenSearch High Level Rest API Client.",
+                    ioException.getMessage());
+            return new OpenSearchHighLevelRestClient(config, bulkListener);
```

Comment:
if `requestInfo(..)` to the elastic search 8.x cluster fails because of a temporary connectivity error we'll end up with an Open Search client that will fail later and will have lots of fun troubleshooting.

I'd just throw an IllegalStateException or similar and fail connector startup.

Reply:
it makes sense! however this is a different behaviour with the current connector because it may fail during the connector creation if the ES sink is temporary down, but I agree it's better to fail at this point